source/common/upstream/upstream_impl.cc

```diff
@@ -228,6 +228,19 @@ absl::optional<uint32_t> HostSetImpl::chooseLocality() {
   return locality->index_;
 }
 
+std::shared_ptr<const HostSet>
+HostSetImpl::filter(std::function<bool(const HostConstSharedPtr&)> predicate) const {
+  HostVector filtered_hosts;
+  HostVector filtered_healthy_hosts;
+
+  std::copy_if(hosts_->begin(), hosts_->end(), std::back_inserter(filtered_hosts), predicate);
+  std::copy_if(healthy_hosts_->begin(), healthy_hosts_->end(),
+               std::back_inserter(filtered_healthy_hosts), predicate);
+
+  return std::make_shared<const HostSetImpl>(priority(), std::move(filtered_hosts),
+                                             std::move(filtered_healthy_hosts));
+}
```

Comment:
This isn't being invoked in the code coverage tests. Is there a way to add a test that will actually trigger this code?

Reply:
Added unit tests. Ideally I'll be able to use this in one of the LB tests (instead of the MockHostSet), I'll play around with it

Reply:
I thought there was already predicate filtering done for subset LB? https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/subset_lb.cc#L535. Is there any common code here that can be shared?